File: C++/Problems/Problems/problem_2952.cpp
#include "problem_2952.h"

bool problem_2952::test() {
    std::vector coins { 1, 4, 10 };
    int target = 19;

    int expected = 2;

    auto result = minimumAddedCoins(coins, target);

    std::cout << result << std::endl;

    return result == expected;
}

int problem_2952::minimumAddedCoins(std::vector<int> &coins, int target) {
    int n = coins.size();
    int result = 0;
    int mx = 0;
    std::sort(coins.begin(), coins.end());
    int i = 0;
    while (mx < target)
    {
        if (i < n && coins[i] <= mx + 1)
        {
            mx += coins[i];
            i++;
        }
        else
        {
            result++;
            mx += mx + 1;
        }
    }

    return result;
}

File: Python/Problems/Problems/problem2952.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem2952(Problem):
    def test(self) -> bool:
        coins = [1, 4, 10]
        target = 19

        expected = 2

        result = self.minimumAddedCoins(coins, target)

        print(result)

        return result == expected

    def minimumAddedCoins(self, coins: List[int], target: int) -> int:
        n = len(coins)
        result = 0
        mx = 0
        coins.sort()
        i = 0
        while (mx < target):
            if i < n and coins[i] <= mx + 1:
                mx += coins[i]
                i += 1
            else:
                result += 1
                mx += mx + 1

        return result